File: backend/jobradar/llm.py
"""LLM access, provider-agnostic.

Gemini is tried first and Groq is the fallback, because their free tiers fail in different
ways: Gemini 2.5 Flash allows far more tokens per minute (~250K vs Groq's ~6-8K), which
matters when the payload is a batch of job descriptions, while Groq's per-day request
allowance is higher. Running out on one provider should not stop the pipeline, so a quota
error on Gemini silently continues on Groq.

Nothing here is on the critical path: if both providers fail, `complete_json` returns None
and enrich.py falls back to heuristics. A run must never fail because of the LLM.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import re
import time
from typing import Any

import httpx

log = logging.getLogger(__name__)

GEMINI_URL = "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent"
GROQ_URL = "https://api.groq.com/openai/v1/chat/completions"

GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-2.5-flash")
GROQ_MODEL = os.getenv("GROQ_MODEL", "llama-3.3-70b-versatile")

# Free-tier request-per-minute ceilings. Staying under these is cheaper than handling the
# 429s they produce.
GEMINI_RPM = 10
GROQ_RPM = 28


class Quota(Exception):
    """Provider refused for quota/rate reasons — try the next provider, do not retry here."""


class _Throttle:
    """Simple RPM limiter shared by all callers of one provider."""

    def __init__(self, rpm: int) -> None:
        self.interval = 60.0 / max(rpm, 1)
        self._lock = asyncio.Lock()
        self._last = 0.0

    async def wait(self) -> None:
        # Checked per call rather than at import so tests (and a paid-tier key, which has
        # much higher limits) can switch throttling off without reloading the module.
        if os.getenv("JOBRADAR_NO_THROTTLE"):
            return
        async with self._lock:
            delta = time.monotonic() - self._last
            if delta < self.interval:
                await asyncio.sleep(self.interval - delta)
            self._last = time.monotonic()


_gemini_throttle = _Throttle(GEMINI_RPM)
_groq_throttle = _Throttle(GROQ_RPM)


def _strip_fence(text: str) -> str:
    """Models wrap JSON in ```json fences even when told not to."""
    text = text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)
    return text.strip()


def available() -> bool:
    return bool(os.getenv("GEMINI_API_KEY") or os.getenv("GROQ_API_KEY"))


async def _gemini(
    client: httpx.AsyncClient, prompt: str, schema: dict[str, Any] | None
) -> str | None:
    key = os.getenv("GEMINI_API_KEY")
    if not key:
        return None

    config: dict[str, Any] = {"temperature": 0.1, "responseMimeType": "application/json"}
    if schema:
        config["responseSchema"] = schema

    await _gemini_throttle.wait()
    resp = await client.post(
        GEMINI_URL.format(model=GEMINI_MODEL),
        headers={"x-goog-api-key": key},
        json={"contents": [{"parts": [{"text": prompt}]}], "generationConfig": config},
        timeout=120.0,
    )
    if resp.status_code in (429, 503):
        raise Quota(f"gemini {resp.status_code}")
    resp.raise_for_status()

    data = resp.json()
    candidates = data.get("candidates") or []
    if not candidates:
        return None
    parts = (candidates[0].get("content") or {}).get("parts") or []
    return "".join(p.get("text", "") for p in parts) or None


async def _groq(client: httpx.AsyncClient, prompt: str, _schema: dict | None) -> str | None:
    key = os.getenv("GROQ_API_KEY")
    if not key:
        return None

    await _groq_throttle.wait()
    resp = await client.post(
        GROQ_URL,
        headers={"Authorization": f"Bearer {key}"},
        json={
            "model": GROQ_MODEL,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": 0.1,
            # Groq has no schema enforcement, only "must be valid JSON".
            "response_format": {"type": "json_object"},
        },
        timeout=120.0,
    )
    if resp.status_code in (429, 503):
        raise Quota(f"groq {resp.status_code}")
    resp.raise_for_status()

    choices = resp.json().get("choices") or []
    return choices[0]["message"]["content"] if choices else None


async def complete_json(
    client: httpx.AsyncClient,
    prompt: str,
    *,
    schema: dict[str, Any] | None = None,
) -> Any | None:
    """Ask for JSON, falling through providers. Returns None if nothing usable came back."""
    for name, fn in (("gemini", _gemini), ("groq", _groq)):
        try:
            raw = await fn(client, prompt, schema)
        except Quota as exc:
            log.info("%s quota exhausted (%s); falling through", name, exc)
            continue
        except Exception as exc:  # noqa: BLE001 - never let a provider fault stop a run
            log.warning("%s call failed: %s", name, exc)
            continue

        if not raw:
            continue
        try:
            return json.loads(_strip_fence(raw))
        except json.JSONDecodeError:
            log.warning("%s returned unparseable JSON", name)
            continue
    return None

File: frontend/src/lib/llm.ts
/**
 * Browser LLM client — Gemini first, Groq as fallback.
 *
 * The key is the viewer's own, kept in localStorage and sent only to Google or Groq. It
 * never reaches this site's server, because there isn't one: the whole app is static files
 * on GitHub Pages. That is what makes the personal features (CV, outreach, interview prep)
 * safe to offer publicly — your CV is never uploaded anywhere we control.
 *
 * Mirrors backend/jobradar/llm.py. Same providers, same fallback order, same JSON contract.
 */

const GEMINI_URL =
  "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent";
const GROQ_URL = "https://api.groq.com/openai/v1/chat/completions";

export const DEFAULT_GEMINI_MODEL = "gemini-2.5-flash";
export const DEFAULT_GROQ_MODEL = "llama-3.3-70b-versatile";

export interface LLMKeys {
  geminiKey?: string;
  groqKey?: string;
  geminiModel?: string;
  groqModel?: string;
}

export class QuotaError extends Error {}
export class NoKeyError extends Error {}

function stripFence(text: string): string {
  const t = text.trim();
  if (!t.startsWith("```")) return t;
  return t.replace(/^```(?:json)?\s*/, "").replace(/\s*```$/, "").trim();
}

export function hasKey(keys: LLMKeys): boolean {
  return Boolean(keys.geminiKey || keys.groqKey);
}

async function callGemini(prompt: string, schema: object | undefined, keys: LLMKeys): Promise<string> {
  const config: Record<string, unknown> = {
    temperature: 0.1,
    responseMimeType: "application/json",
  };
  if (schema) config.responseSchema = schema;

  const resp = await fetch(
    GEMINI_URL.replace("{model}", keys.geminiModel || DEFAULT_GEMINI_MODEL),
    {
      method: "POST",
      headers: { "Content-Type": "application/json", "x-goog-api-key": keys.geminiKey! },
      body: JSON.stringify({
        contents: [{ parts: [{ text: prompt }] }],
        generationConfig: config,
      }),
    },
  );

  if (resp.status === 429 || resp.status === 503) throw new QuotaError("Gemini rate limit");
  if (resp.status === 400 || resp.status === 403) {
    throw new Error("Gemini rejected the key. Check it at aistudio.google.com/apikey");
  }
  if (!resp.ok) throw new Error(`Gemini returned ${resp.status}`);

  const data = await resp.json();
  const parts = data?.candidates?.[0]?.content?.parts ?? [];
  const text = parts.map((p: { text?: string }) => p.text ?? "").join("");
  if (!text) throw new Error("Gemini returned an empty response");
  return text;
}

async function callGroq(prompt: string, keys: LLMKeys): Promise<string> {
  const resp = await fetch(GROQ_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${keys.groqKey}`,
    },
    body: JSON.stringify({
      model: keys.groqModel || DEFAULT_GROQ_MODEL,
      messages: [{ role: "user", content: prompt }],
      temperature: 0.1,
      response_format: { type: "json_object" },
    }),
  });

  if (resp.status === 429 || resp.status === 503) throw new QuotaError("Groq rate limit");
  if (resp.status === 401) throw new Error("Groq rejected the key. Check it at console.groq.com");
  if (!resp.ok) throw new Error(`Groq returned ${resp.status}`);

  const data = await resp.json();
  const text = data?.choices?.[0]?.message?.content;
  if (!text) throw new Error("Groq returned an empty response");
  return text;
}

/** Ask for JSON, falling through providers. Throws with a message worth showing a user. */
export async function completeJSON<T = unknown>(
  prompt: string,
  keys: LLMKeys,
  schema?: object,
): Promise<T> {
  if (!hasKey(keys)) {
    throw new NoKeyError("Add a Gemini or Groq API key in Settings. Both have free tiers.");
  }

  const errors: string[] = [];

  for (const provider of ["gemini", "groq"] as const) {
    if (provider === "gemini" && !keys.geminiKey) continue;
    if (provider === "groq" && !keys.groqKey) continue;

    try {
      const raw =
        provider === "gemini"
          ? await callGemini(prompt, schema, keys)
          : await callGroq(prompt, keys);
      return JSON.parse(stripFence(raw)) as T;
    } catch (err) {
      if (err instanceof QuotaError) {
        // Falling through to the other provider is the whole point of holding two keys.
        errors.push(`${provider}: quota exhausted`);
        continue;
      }
      if (err instanceof SyntaxError) {
        errors.push(`${provider}: returned unparseable JSON`);
        continue;
      }
      errors.push(`${provider}: ${err instanceof Error ? err.message : String(err)}`);
    }
  }

  throw new Error(errors.join(" · ") || "No provider produced a usable response.");
}
